game/songparser.hh

```diff
@@ -10,105 +10,103 @@
-		bool getline (std::string& line) { ++m_linenum; return (bool) std::getline (m_ss, line); }
-		bool m_relative;
-		double m_gap;
-		double m_bpm;
-
-		bool txtCheck(std::string const& data) const;
-		void txtParseHeader();
-		void txtParse();
-		bool txtParseField(std::string const& line);
-		bool txtParseNote(std::string line);
-		bool iniCheck(std::string const& data) const;
-		void iniParseHeader();
-		bool midCheck(std::string const& data) const;
-		void midParseHeader();
-		void midParse();
-		bool xmlCheck(std::string const& data) const;
-		void xmlParseHeader();
-		void xmlParse();
-		Note xmlParseNote(xmlpp::Element const& noteNode, unsigned& ts);
-		bool smCheck(std::string const& data) const;
-		void smParseHeader();
-		void smParse();
-		bool smParseField(std::string line);
-		Notes smParseNotes(std::string line);
-		void resetNoteParsingState();
-		double m_prevtime;
-		unsigned int m_prevts;
-		unsigned int m_relativeShift;
-		enum CurrentSinger { P1 = 1, P2 = 2, BOTH = P1 | P2 } m_curSinger;
-		struct BPM {
-			BPM (double _begin, double _ts, double bpm) :
-				begin (_begin), step (0.25 * 60.0 / bpm), ts (_ts) {}
-			double begin;	// Time in seconds
-			double step;// Seconds per quarter note
-			double ts;
-			};
-		typedef std::vector<BPM> bpms_t;
-		bpms_t m_bpms;
-		unsigned m_tsPerBeat;	// /< The ts increment per beat
-		unsigned m_tsEnd;	// /< The ending ts of the song
-		BPM getBPM (double ts) {
-			for (auto& itb : boost::adaptors::reverse (m_bpms)) {
-				if (itb.begin > ts) { 
-					continue; 
-				} else {
-					return itb;
-				}
-			}
-			throw std::runtime_error("No BPM definition prior to this note...");
+public:
+	/// Parse into s
+	SongParser (Song & s);
+private:
+	// Variables and types
+	Song& m_song;
+	std::stringstream m_ss;
+	unsigned m_linenum = 0;
+	bool m_relative = false;
+	double m_gap = 0.0;
+	double m_bpm = 0.0;
+	unsigned m_tsPerBeat = 0;  ///< The ts increment per beat
+	unsigned m_tsEnd = 0;  ///< The ending ts of the song
+	enum class CurrentSinger { P1, P2, BOTH } m_curSinger = CurrentSinger::P1;
+	struct BPM {
+		BPM (double _begin, double _ts, double bpm) :
+		begin (_begin), step (0.25 * 60.0 / bpm), ts (_ts) {}
+		double begin;  // Time in seconds
+		double step;  // Seconds per quarter note
+		double ts;
+	};
+	std::vector<BPM> m_bpms;
+	Song::Stops m_stops;  ///< Stops stored in <ts, duration> format
+	/// The following struct is cleared between tracks
+	struct TXTState {
+		double prevtime = 0.0;
+		unsigned prevts = 0;
+		unsigned relativeShift = 0;
+	} m_txt;
+	// Functions
+	void finalize();
+	void vocalsTogether();
+	void guessFiles();
+	bool getline (std::string& line) { ++m_linenum; return (bool) std::getline (m_ss, line); }
+	BPM getBPM (double ts) {
+		for (auto& itb : boost::adaptors::reverse (m_bpms)) {
+			if (itb.begin <= ts) return itb;
+		}
+		throw std::runtime_error("No BPM definition prior to this note...");
+	}
+	void addBPM (double ts, double bpm) {
```

Comment:
please move functions out of the header.

Reply:
Moved now.